src/core/shelf.sol

```diff
@@ -46,40 +42,63 @@ contract Shelf is TitleOwned {
     modifier auth { require(wards[msg.sender] == 1); _; }
 
     // --- Data ---
-    PileLike                  public pile;
-    AppraiserLike             public appraiser;
+    PileLike public  pile;
+    Title public     title;
+    DebtLike public  debt;
 
     struct Loan {
         address registry;
         uint256 tokenId;
-        uint price; //
-        uint principal; //
-        uint initial; //
+        uint price;
+        uint principal;
+        uint initial;
     }
 
-    mapping (uint => Loan) public shelf;
+    mapping (uint => Loan) public    shelf;
+    mapping (bytes32 => uint) public nftlookup;
 
     uint public bags; // sum of all prices
 
-    constructor(address pile_, address appraiser_, address title_) TitleOwned(title_) public {
+    constructor(address pile_, address title_) TitleOwned(title_) public {
         wards[msg.sender] = 1;
         pile = PileLike(pile_);
-        appraiser = AppraiserLike(appraiser_);
+        title = Title(title_);
     }
 
-    // --- Shelf ---
-
     function depend (bytes32 what, address addr) public auth {
         if (what == "pile") { pile = PileLike(addr); }
-        else if (what == "appraiser") { appraiser = AppraiserLike(addr); }
         else revert();
     }
 
-    function file(uint loan, address registry_, uint nft_) public auth {
-        shelf[loan].registry = registry_;
-        shelf[loan].tokenId = nft_;
+    // --- Shelf: Getters ---
+    function token(uint loan) public view returns (address registry, uint nft) {
+        return (shelf[loan].registry, shelf[loan].tokenId);
+    }
+
+    // --- Shelf: Creation and closing of a loan ---
+    function issue(address registry, uint token) public returns (uint) {
+        require(NFTLike(registry).ownerOf(token) == msg.sender, "nft-not-owned");
+
+        bytes32 nft = keccak256(abi.encodePacked(registry, token));
+        require(nftlookup[nft] == 0, "nft-in-use");
+
+        uint loan = title.issue(msg.sender);
+        nftlookup[nft] = loan;
+
+        shelf[loan].registry = registry;
+        shelf[loan].tokenId = token;
+
+        return loan;
     }
 
+    function close(uint loan) public owner(loan) {
+        require(pile.debtOf(loan) == 0, "outstanding-debt"); // TODO: only allow closing of a loan that isn't active anymore. maybe there is a better criteria
```

Comment:
in case of async borrowing the loan might have a balance that has not been withdrawn yet. If we don't decrease it, we might end up taking more money from the lending side then required.

Reply:
yeah, I would like to discuss the balance handling again, I think there are a few of these issues still in the code base.